### PR TITLE
Refactor lti launches to be more explicit about assignments 

### PR DESCRIPTION
**Pull request by marcospri**
This PR doesn't contain any D2L deep linking code but it's motivated by that.

Being more explicit about what we are doing with assignments and when will allow to plug a small difference for D2L (to handle groups).

I haven't explored here the idea we discussed about a potential "AssignmentConfiguration` dataclass or similar to keep things more focused now but I reckon is more evident in this PR where that would go.

The main change here are:

- replacing `upsert_assignment` with more explicit methods and a `get_assignment_for_launch`
- Add more parameters to the plugins that get document_url and group_set_id

## Testing

Some sanity checking Canvas and D2L, no new behaviors here.

### Canvas

- Use a teacher login, eg: `eng+canvasteacher@hypothes.is`

- [Reconfigure a canvas assignment](https://hypothesis.instructure.com/courses/125/assignments/4682/edit?return_to=https%3A%2F%2Fhypothesis.instructure.com%2Fcourses%2F125%2Fassignments%2F4682)

- Pick any document source, and groups
- Launch with that document on groups
- [Reconfigure a canvas assignment](https://hypothesis.instructure.com/courses/125/assignments/4682/edit?return_to=https%3A%2F%2Fhypothesis.instructure.com%2Fcourses%2F125%2Fassignments%2F4682)
 again
- Pick a different document source and disable groups
- Launch it and you'll get the new document and a sections drop down.

### D2L 

Use a teacher login, eg `HypothesisEng.Teacher`

- Truncate assignments to be able to use the scratch one 

```docker compose exec postgres psql -U postgres -c "truncate assignment cascade"```

- Launch and configure a [D2L assignment](https://aunltd.brightspacedemo.com/d2l/le/content/6782/viewContent/2431/View)
- Pick one document and a group set
- You'll get the document and group selection
- Launch it again from scratch D2L assignment](https://aunltd.brightspacedemo.com/d2l/le/content/6782/viewContent/2431/View)
- Click edit, select a different D2L document and disable groups.
- You'll get the new document and the course group.

### CODE REVIEW

File: lms/product/plugin/misc.py [hunk 1]
```diff
@@ -40,24 +40,10 @@ def get_ltia_aud_claim(self, lti_registration: LTIRegistration) -> str:
         """Get the value of the `aud` claim used in LTI advantage requests."""
         return lti_registration.token_url
 
-    def get_document_url(self, request):
+    def get_document_url(self, _request, assignment, historical_assignment):
```
Comment (marcospri) [inline lms/product/plugin/misc.py:43]: In the general case get if from the assignment or the copy.

Not sorter code but note that we remove one place to query for the copy and assignment. I reckon having only place doing that is much easier to follow.

File: lms/services/assignment.py [hunk 2]
```diff
@@ -56,59 +71,49 @@ def get_copied_from_assignment(self, lti_params) -> Optional[Assignment]:
...
-        is_gradable=False,
-        extra=None,
-    ) -> Assignment:
-        """
-        Update or create an assignment with the given document_url.
-
-        Set the document_url for the assignment that matches
-        tool_consumer_instance_guid and resource_link_id
-        or create a new one if none exist on the DB.
+    def get_assignment_for_launch(self, request):
```
Comment (marcospri) [inline lms/services/assignment.py:74]: `get_or_create`?

Comment (marcospri) [inline lms/services/assignment.py:74]: I generally don't like very broad parameters like `request` but in this case, is a `here the request, give me an assignment`.

The rest of the methods in the servie have better scoped parameters.

File: lms/services/assignment.py [hunk 3]
```diff
@@ -56,59 +71,49 @@ def get_copied_from_assignment(self, lti_params) -> Optional[Assignment]:
...
+        # Always update the assignment ULR
         assignment.document_url = document_url
+
+        # And metadata based on the launch
         assignment.title = lti_params.get("resource_link_title")
         assignment.description = lti_params.get("resource_link_description")
-        assignment.is_gradable = is_gradable
-        if extra:
-            assignment.extra = extra
+        assignment.is_gradable = self._misc_plugin.is_assignment_gradable(lti_params)
```
Comment (marcospri) [inline lms/services/assignment.py:116]: Have to think about this. In the current state of this PR this are not updated for new and edited assignments, only on launches.

File: tests/unit/conftest.py [hunk 4]
```diff
@@ -124,7 +124,7 @@ def pyramid_request(db_session, application_instance, lti_v11_params):
     )
 
     pyramid_request.lti_jwt = {}
-    pyramid_request.lti_params = {}
+    pyramid_request.lti_params = LTIParams(v11=lti_v11_params)
```
Comment (marcospri) [inline tests/unit/conftest.py:127]: Missing form this fixture

File: tests/unit/lms/product/factory_test.py [hunk 5]
```diff
@@ -125,4 +125,5 @@ def with_mocked_from_request(self):
     def pyramid_request(self, pyramid_request):
         pyramid_request.matched_route = Mock(spec_set=["name"])
         pyramid_request.matched_route.name = "some.route"
+        pyramid_request.lti_params = {}
```
Comment (marcospri) [inline tests/unit/lms/product/factory_test.py:128]: Fixed the fixture to have values for this.

Here we were not expecting them so needs fixing.

File: lms/views/lti/basic_launch.py [hunk 6]
```diff
@@ -148,36 +155,6 @@ def _show_document(self, document_url, assignment_extra=None):
...
-
-        assignment_gradable = self.request.product.plugin.misc.is_assignment_gradable(
-            self.request.lti_params
-        )
-
-        # Store lots of info
-        assignment = self._record_assignment(
-            document_url, extra=assignment_extra, is_gradable=assignment_gradable
-        )
-
```
Comment (marcospri) [inline lms/views/lti/basic_launch.py:162]: Inlined this.

A while ago we have a few things here but now it's just "_record_assignment_relationships" 

If we go back to "store lots" it's easy enough to add a private method.

**Comment by marcospri** [comment]
> Why aren't configure_assignment_callback and edit_assignment_callback the same? One calls create not get + we have notify in one

Might explore this. Leaving it for now

> Should post_configure_assignment move to update_assignment ? Yes, but we need to be careful it only gets called when it should. Could use a private update_assignment for not calling the plugin.

Done exactly this :+1: 


> Show document doesn't seem to do much showing of documents.
> Two calls here might be better than one to unwind this into separate pieces of code, instead of one long call chain

Left configure_and_show_document as two method calls from the callers.

show_document needs a better name and probably there's better opportunities to make the whole file saner now in follow ups.

I have this PR about some conditionals further down in the file: https://github.com/hypothesis/lms/pull/5469


> We could accept lti_params for some of the getting / updating / creating methods and cut down on repetition for callers getting certain params

Tried this but IMO didn't improve that much and we use at least the get version with different arguments than the ones LTIParams.


> _misc_plugin.get_document_url() + _grouping_plugin.get_group_set_id() -> _assignment_plugin.get_assignment_config() ?
> Maybe, but it's not the motivation for the current code layout.
> It's because we need to get this before we potentially overwrite it

`get_assignment_config` will be something to explore.

> Is the end of get_assignment_for_launch actually what update_assignment should be?
> Yes, and it might be a bug to not update title etc
> Need to be careful with the post call plugin (if that's not also a bug?)

Move parts of get_assignment_for_launch to update_assignment.

> Is get_copied_from_assignment private now? Yes

:+1:

File: lms/views/lti/basic_launch.py [hunk 7]
```diff
@@ -35,12 +33,10 @@ def __init__(self, context, request):
         self.assignment_service: AssignmentService = request.find_service(
             name="assignment"
         )
-        self.grouping_service: GroupingService = request.find_service(name="grouping")
-        self.course_service: CourseService = request.find_service(name="course")
```
Comment (marcospri) [inline lms/views/lti/basic_launch.py:39]: This are only used on record_course

File: lms/views/lti/basic_launch.py [hunk 8]
```diff
@@ -35,12 +33,10 @@ def __init__(self, context, request):
...
+        self._guid = self.request.lti_params.get("tool_consumer_instance_guid")
+        self._resource_link_id = self.request.lti_params.get("resource_link_id")
```
Comment (marcospri) [inline lms/views/lti/basic_launch.py:37]: Maybe a bit silly to do this, avoid some typing :shrug: 

I was thinking that on top of the dict interface (maybe eventually replacing it?) we could dataclass-isfy LTIParams to do

`request.lti_params.guid`


Not for this PR.

File: lms/product/plugin/grouping.py [hunk 9]
```diff
@@ -74,18 +74,27 @@ def sections_enabled(self, request, application_instance, course) -> bool:
...
+        `historical_assignment` might be none even when it might exist on the DB
+        (or even available as assignment.copied_form).
+        The historical_assignment is only relevant to get the config when creating an assignment for the first time.
         """
-        if not self.group_type or not assignment:
+        if not self.group_type:
+            # Groups not enabled on this product
             return None
 
+        if not assignment and historical_assignment:
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply lms/product/plugin/grouping.py:94]: I've been staring at this and I'm not sure I quite grok it. Is it this?

```python
        if assignment:
            return assignment.extra.get("group_set_id")

        if historical_assignment:
            # When creating new assignments, take the value from the previous
            # version of the assignment
            return historical_assignment.extra.get("group_set_id")

        return None
```

Because I think that's the same logic as in the `misc.py:get_document_url`.

File: lms/product/plugin/misc.py [hunk 10]
```diff
@@ -40,27 +40,13 @@ def get_ltia_aud_claim(self, lti_registration: LTIRegistration) -> str:
...
-            resource_link_id=request.lti_params.get("resource_link_id"),
-        )
-
-        if not assignment:
-            # If the current assignment is not yet in the DB maybe we
-            # are launching for the first time a copied assignment.
-            assignment = assignment_service.get_copied_from_assignment(
-                request.lti_params
-            )
-
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply lms/product/plugin/misc.py:62]: This is much nicer. I wonder if we could skip the trinary at the bottom to make it the same as the suggestion above for the grouping plugin?

```python
        if assignment:
            return assignment.document_url

        if historical_assignment:
            return historical_assignment.document_url

        return None
```

I find this a bit easier to read.

File: lms/services/assignment.py [hunk 11]
```diff
@@ -56,61 +91,58 @@ def get_copied_from_assignment(self, lti_params) -> Optional[Assignment]:
...
+            # We don't have an assignment in the DB but we know to which document url it should point
+            # This might happen for example on:
+            #   - the first launch of a deep linked assignment
+            #   - the first launch copied assignment
+            assignment = self.create_assignment(
+                tool_consumer_instance_guid, resource_link_id
+            )
 
-        return assignment
+            if historical_assignment:
```
Comment (jon-betts) [inline lms/services/assignment.py:135]: Do we need to if here? If historical assignment is None, presumably it would have no effect?

**Review by jon-betts — APPROVED** [review]
I've followed through the testing notes and they work. I've not gone through the DB with a fine tooth comb.

I'm not sure there's anyway, or if I'm just a bit dim this week, but I find this PR hard to follow along with all at once. I wonder if we could have changed a few things in smaller chunks? Perhaps not.

As far as suggestions go, there are lots of nit picky suggestions about comments. I still think we should stick to the PEP-8 79 limit. Aside from that, actual concrete suggestions:

* Having an underscore in the name of a variable in a base class is weird, I think that should change, as it forces us to never use kwargs or have weird kwargs
* As mentioned in chat, there are many follow up improvements, I think this is definitely "not done"
* Some of the if logic seems convoluted in places?
* Private methods should go after public ones
  * In the case of `_configure_js_to_show_document` it's only called right at the end of `_show_document` and it's miles away. Would be nice to put it direclty underneith so you can read from one to the next. In general have a bit of a think to see if readability can be enhanced with ordering.
